src/coreclr/jit/emitarm64.h

```diff
@@ -643,7 +646,16 @@ static code_t insEncodeSplitUimm(size_t imm)
     size_t immhi = (imm >> lo_bits) & (hi_max - 1);
     size_t immlo = imm & (lo_max - 1);
 
-    return insEncodeUimm<hi1, lo1>(immhi) | insEncodeUimm<hi2, lo2>(immlo);
+    code_t result = insEncodeUimm<hi1, lo1>(immhi) | insEncodeUimm<hi2, lo2>(immlo);
+
+    // Calculate and generate a mask for the number of bits between hi2-lo1, and assert that these bits
+    // are not set in the result. Note if between_bits == 0 then the mask will always be 0 and this will
+    // pass.
+    size_t between_bits = lo1 - hi2 - 1;
+    code_t between_mask = ((1 << between_bits) - 1) << (hi2 + 1);
+    assert((result & between_mask) == 0);
```

Comment:
I may have misunderstood your comment, did you mean that it shouldn't be allowed to do `insEncodeSplitUimm<23, 20, 20, 16>` (because that should be `insEncodeUimm<23, 16>`)?

Reply:
I'm not sure I can check this early because the immediate isn't encoded yet when it is passed into this function? For example you would be calling this with some value like an index '5' and it will convert 5 => 0000101 and then split this and encode it into `0b00 << 21 | 0b00101 << 16`. But I think it is worth asserting this end value doesn't have any wrong bits encoded in it because the bitwise logic is a bit complicated.